### PR TITLE
[GH-#269] Replace sha256 with blake2 for Oracles

### LINKED ISSUE #269: replace sha256 with blake2 as default hash function
research where blake2 is used in the core, other than oracles https://www.pivotaltracker.com/epic/show/3860576, document findings here

do after #212 is merged

### PR DESCRIPTION
**Pull request by HristianHristov**
Resolves #269

### CODE REVIEW

**Comment by HristianHristov** [comment]
Sorry for missunderstanding, but the issue name was for the oracles: https://github.com/aeternity/elixir-node/issues/269
Will be done soon.

**Comment by thepiwo** [comment]
oh, then it was my misunderstanding, but as oracles are merged, we can now adjust it for everything

File: necessary.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+curl -fsSL -o libsodium-src.tar.gz "https://github.com/jedisct1/libsodium/releases/download/1.0.16/libsodium-1.0.16.tar.gz"
+mkdir libsodium-src
+tar -zxf libsodium-src.tar.gz -C libsodium-src --strip-components=1
+cd libsodium-src/
+sudo ./configure &&  sudo make -j$(nproc) &&  sudo make install && sudo ldconfig
```
Comment (cytadela8) [inline necessary.sh]: Do we really need to download libsodium sources? Isn't there an ubuntu package with this lib?

Comment (cytadela8) [reply necessary.sh]: Also: you shouldn't run `make -j$(nproc)` as root

Comment (cytadela8) [reply necessary.sh]: If we really need to download and compile libsodium then maybe we can avoid installing it to system? You already use `LD_LIBRARY_PATH` to some path. This configured properly should be enough.

Comment (thepiwo) [reply necessary.sh]: compilation is needed, as repos are outdated

Comment (thepiwo) [reply necessary.sh]: but this file is not needed, a note in the readme is sufficient I think

Comment (thepiwo) [reply necessary.sh]: I did this in the `Dockerfile`

Comment (HristianHristov) [reply necessary.sh]: Ok I will make a note in the readme and remove this script

File: README.md [hunk 2]
```diff
@@ -27,6 +27,9 @@ Make sure you have installed the following packages to make sure that the Wallet
 sudo apt-get install autoconf autogen
 sudo apt-get install libtool
 sudo apt-get install libgmp3-dev
+wget -O libsodium-src.tar.gz https://github.com/jedisct1/libsodium/releases/download/1.0.16/libsodium-1.0.16.tar.gz
+ mkdir libsodium-src && tar -zxf libsodium-src.tar.gz -C libsodium-src --strip-components=1
+  cd libsodium-src && ./configure --prefix=$HOME/.libsodium && make -j$(nproc) && make install && cd ..
```
Comment (thepiwo) [inline README.md]: installation must be sudo

File: apps/aeutil/lib/hash.ex [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    :sha3.hash(@hash_bytes_size * 8, data)
+  end
+
+  @spec hash_sha256(binary()) :: binary()
+  def hash_sha256(data) when is_binary(data) do
+    :crypto.hash(:sha256, data)
+  end
+
+  @spec hash(binary()) :: binary()
+  defdelegate hash(opts), to: __MODULE__, as: :hash_blake2b
```
Comment (d-velev) [inline apps/aeutil/lib/hash.ex]: If this exists then Hash.hash_blake2b should be replaced with Hash.hash right?

Comment (d-velev) [reply apps/aeutil/lib/hash.ex]: this was commented on, didn't see

**Comment by thepiwo** [comment]
@HristianHristov what is the state on this?

**Comment by HristianHristov** [comment]
@thepiwo Currently I'm trying to fix the Cuckoo tests, but for some reason I can't use the generichash function, because it couldn't find the :enacl_nif.so

**Comment by thepiwo** [comment]
@d-velev @cytadela8 can you re-review?

**Pull request merged**